here-naksha-app-service/src/test/java/com/here/naksha/app/service/ValDryRunTest.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.app.service;
+
+import static com.here.naksha.app.common.CommonApiTestSetup.*;
+import static com.here.naksha.app.common.ResponseAssertions.assertThat;
+import static com.here.naksha.app.common.TestUtil.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+import com.here.naksha.app.common.ApiTest;
+import com.here.naksha.app.common.NakshaTestWebClient;
+import com.here.naksha.app.service.models.FeatureCollectionRequest;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzFeature;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzFeatureCollection;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzReference;
+import com.here.naksha.lib.core.models.naksha.Space;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.UUID;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONException;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.skyscreamer.jsonassert.JSONAssert;
+import org.skyscreamer.jsonassert.JSONCompareMode;
+import org.skyscreamer.jsonassert.comparator.ArraySizeComparator;
+
+public class ValDryRunTest extends ApiTest {
+
+  private static final NakshaTestWebClient nakshaClient = new NakshaTestWebClient();
+
+  private static final String SPACE_ID = "local-space-4-val-dry-run";
+
+  @BeforeAll
+  static void setup() throws URISyntaxException, IOException, InterruptedException {
+    createHandler(nakshaClient, "ValDryRun/setup");
+    createSpace(nakshaClient, "ValDryRun/setup");
+  }
+
+  private void additionalCustomAssertions_tc3000(final @NotNull String reqBody, final @NotNull String resBody)
+      throws JSONException {
+    final FeatureCollectionRequest collectionRequest = parseJson(reqBody, FeatureCollectionRequest.class);
+    final XyzFeatureCollection collectionResponse = parseJson(resBody, XyzFeatureCollection.class);
+    final List<String> updatedIds = collectionResponse.getUpdated();
+    final List<XyzFeature> features = collectionResponse.getFeatures();
+    final List<XyzFeature> violations = collectionResponse.getViolations();
+    JSONAssert.assertEquals(
+        "{updated:[" + collectionRequest.getFeatures().size() + "]}",
+        resBody,
+        new ArraySizeComparator(JSONCompareMode.LENIENT));
+    assertEquals(
+        updatedIds.size(), features.size(), "Mismatch between updated and features list size in the response");
+    final String newFeatureId = features.get(2).getId();
```

Comment:
Why do check only third element? This seems to be case-specific but the method is available for all of test cases.
Similarly: why do we iterate over 4..6th elements in the for loop? This also implies that this method is for specific case, rather than common helper method

Reply:
For first question - As the function name suggests, it is specific to a test case (tc3000). Yes, theoretically speaking, it is possible other tests can also call it and may fail unknowingly. May be once refactoring is done, this can be improved.
W.r.t. other question - we only need to validate violation IDs of those features whose ID is not known while making the request (as the ID is generated in the response itself). Hence, only specific features/violation IDs are matched. Others are already getting compared during standardAssertion of entire response body.